src/main/java/io/divolte/server/kafka/KafkaFlusher.java

```diff
@@ -81,8 +86,9 @@ public ProcessingDirective process(final Queue<AvroRecordBuffer> batch) {
             logger.debug("Processing batch of {} records.", batchSize);
             final List<ProducerRecord<DivolteIdentifier, AvroRecordBuffer>> kafkaMessages =
                     batch.stream()
+                         .map(i -> i.payload)
                          .map(this::buildRecord)
-                         .collect(Collectors.toList());
+                         .collect(Collectors.toCollection(() -> new ArrayList<>(batchSize)));
```

Comment:
Much of the mechanics in here feel a bit allocation-heavy. I don't know the details of the approach, so cannot suggest an alternative right now, but perhaps something to look into.